src/modules/flow/twitter/twitter.c

```diff
@@ -0,0 +1,617 @@
+    SOL_INT_CHECK_GOTO(r, < 0, err_escape_status);
+    escaped_status = sol_str_slice_to_string(sol_buffer_get_slice(&buf));
+    sol_buffer_fini(&buf);
+    SOL_NULL_CHECK_GOTO(escaped_status, err_escape_status);
+
+    digest_config->data = cb_data;
+    digest_config->key = sol_str_slice_from_str(cb_data->key);
+
+    digest = sol_message_digest_new(digest_config);
+    SOL_NULL_CHECK_GOTO(digest, err_digest);
+
+    r  = asprintf(&params,
+        "oauth_consumer_key=%s"
+        "&oauth_nonce=%s"
+        "&oauth_signature_method=HMAC-SHA1"
+        "&oauth_timestamp=%s"
+        "&oauth_token=%s"
+        "&oauth_version=1.0"
+        "&status=%s",
+        mdata->consumer_key, cb_data->nonce,
+        cb_data->timestamp, mdata->token, escaped_status);
+    SOL_INT_CHECK_GOTO(r, < 0, err_params);
+
+    r = sol_http_encode_slice(&buf, sol_str_slice_from_str(params));
+    SOL_INT_CHECK_GOTO(r, < 0, err_escape_params);
+    escaped_params = sol_str_slice_to_string(sol_buffer_get_slice(&buf));
+    sol_buffer_fini(&buf);
+    SOL_NULL_CHECK_GOTO(escaped_params, err_escape_params);
+
+    r = asprintf(&signature, "POST&%s&%s", mdata->escaped_post_url, escaped_params);
+    SOL_INT_CHECK_GOTO(r, < 0, err_signature);
+
+    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, signature, strlen(signature));
+    SOL_NULL_CHECK_GOTO(blob, err_blob);
+
+    r = sol_message_digest_feed(digest, blob, true);
+    SOL_INT_CHECK_GOTO(r, < 0, err_feed);
+
+    sol_blob_unref(blob);
+    free(escaped_status);
+    free(escaped_params);
+    free(params);
+
+    return 0;
+
+err_feed:
+    sol_blob_unref(blob);
+err_blob:
+    free(signature);
+err_signature:
+    free(escaped_params);
+err_escape_params:
+    free(params);
+err_params:
+    sol_message_digest_del(digest);
+err_digest:
+    free(escaped_status);
+err_escape_status:
+    free(cb_data->status);
+err_status:
+    free(cb_data->key);
+err_key:
+    free(cb_data->timestamp);
+err_timestamp:
+    free(cb_data->nonce);
+err:
+    free(cb_data);
+    return r;
+}
+
+static int
+token_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    int r;
+    uint16_t idx;
+    const char *val;
+    size_t token_len, token_secret_len;
+    struct sol_buffer buf;
+    struct sol_vector tokens;
+    struct sol_str_slice *slice;
+    struct twitter_data *mdata = data;
+
+    free(mdata->token);
+    mdata->token = NULL;
+    free(mdata->token_secret);
+    mdata->token_secret = NULL;
+
+    r = sol_flow_packet_get_string(packet, &val);
+    SOL_INT_CHECK(r, < 0, r);
+
+    token_len = sizeof("oauth_token=") - 1;
+    token_secret_len = sizeof("oauth_token_secret=") - 1;
+
+    tokens = sol_util_str_split(
+        sol_str_slice_from_str(val), "&", 0);
+    SOL_VECTOR_FOREACH_IDX (&tokens, slice, idx) {
+        struct sol_str_slice token;
+        if (streqn(slice->data, "oauth_token_secret=", token_secret_len) &&
```

Comment:
use the new function that @cabelitos added slice-starts with.

@cabelitos, please review this commit as you added lots of new features that could improve it :-)